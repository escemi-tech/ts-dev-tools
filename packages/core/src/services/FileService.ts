import { copyFileSync, existsSync, readFileSync, writeFileSync } from "fs";

export class FileService {
  static fileExists(filePath: string): boolean {
    return existsSync(filePath);
  }

  static getFileContent(filePath: string): string {
    return readFileSync(filePath, "utf-8");
  }

  static putFileContent(path: string, content: string) {
    writeFileSync(path, content);
  }

  static copyFile(sourcePath: string, destinationPath: string) {
    copyFileSync(sourcePath, destinationPath);
  }
}
