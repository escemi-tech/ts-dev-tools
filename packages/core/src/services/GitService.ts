import { chmodSync, createWriteStream, existsSync, statSync } from "fs";
import { join } from "path";

import { PROJECT_NAME, PROJECT_URL } from "../constants";
import { CmdService } from "./CmdService";

export class GitService {
  static GIT_HOOK_MODE = 0o755;
  static GIT_HOOK_TEMPLATE = `#!/bin/sh

# Created by ${PROJECT_NAME} (${PROJECT_URL})

%gitHookCommand%`;

  static async isGitRepository(absoluteProjectDir: string): Promise<boolean> {
    return await CmdService.execCmd("git rev-parse", absoluteProjectDir, true)
      .then(() => true)
      .catch(() => false);
  }

  static async addGitHook(
    absoluteProjectDir: string,
    gitHookName: string,
    gitHookCommand: string
  ): Promise<void> {
    const gitHookDirPath = join(absoluteProjectDir, ".git/hooks");
    const gitHookFilePath = join(gitHookDirPath, gitHookName);

    if (existsSync(gitHookFilePath)) {
      const mode = GitService.getFilePermissions(gitHookFilePath);
      if (mode !== GitService.GIT_HOOK_MODE) {
        chmodSync(gitHookFilePath, GitService.GIT_HOOK_MODE);
      }
      return;
    }

    return new Promise((resolve, reject) => {
      const stream = createWriteStream(gitHookFilePath, { mode: GitService.GIT_HOOK_MODE });
      stream.on("close", () => resolve(undefined));
      stream.on("error", (error) => reject(error));
      stream.write(GitService.GIT_HOOK_TEMPLATE.replace("%gitHookCommand%", gitHookCommand));
      stream.end();
    });
  }

  private static getFilePermissions(filePath: string) {
    const { mode } = statSync(filePath);
    return mode & parseInt("777", 8);
  }
}
