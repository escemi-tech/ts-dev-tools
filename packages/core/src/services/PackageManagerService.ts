import { spawn } from "child_process";
import { existsSync } from "fs";
import { join } from "path";

export enum PackageManagerType {
  yarn = "yarn",
  npm = "npm",
}

export class PackageManagerService {
  static detectPackageManager(dirPath: string): PackageManagerType {
    if (existsSync(join(dirPath, "yarn.lock"))) {
      return PackageManagerType.yarn;
    }
    return PackageManagerType.npm;
  }

  static async addDevPackage(packageName: string, dirPath: string): Promise<void> {
    const packageManager = PackageManagerService.detectPackageManager(dirPath);
    const isMonorepo = await PackageManagerService.isMonorepo(dirPath);

    const args: string[] = [packageManager];

    switch (packageManager) {
      case PackageManagerType.yarn:
        args.push("add", "--dev");

        if (isMonorepo) {
          args.push("--ignore-workspace-root-check");
        }

        break;
      case PackageManagerType.npm:
        args.push("install", "--save-dev");

        if (isMonorepo) {
          args.push("--no-workspaces");
        }
        break;
    }

    args.push(packageName);

    await PackageManagerService.execCommand(args, dirPath, true);
  }

  static async isMonorepo(dirPath: string) {
    const packageManager = PackageManagerService.detectPackageManager(dirPath);

    const args: string[] = [packageManager];

    switch (packageManager) {
      case PackageManagerType.yarn:
        args.push("workspaces", "info");
        break;

      case PackageManagerType.npm:
        args.push("--workspaces", "list");
        break;
    }

    args.push("> /dev/null 2>&1 && echo true || echo false;");

    const output = await PackageManagerService.execCommand(args, dirPath, true);

    return output.trim() === "true";
  }

  static async isPackageInstalled(packageName: string, dirPath: string): Promise<boolean> {
    const packageManager = PackageManagerService.detectPackageManager(dirPath);

    const args = [
      packageManager,
      "list",
      "--depth=1",
      "--json",
      "--no-progress",
      `--pattern="${packageName}"`,
      "--non-interactive",
    ];

    const output = await PackageManagerService.execCommand(args, dirPath, true);

    const installedPackages = JSON.parse(output);

    switch (packageManager) {
      case PackageManagerType.yarn:
        return installedPackages?.data?.trees?.some((tree: { name: string }) =>
          tree.name.startsWith(packageName + "@")
        );
      case PackageManagerType.npm:
        return installedPackages.dependencies
          ? Object.prototype.hasOwnProperty.call(installedPackages.dependencies, packageName)
          : false;
    }
  }

  private static async execCommand(
    args: string | string[],
    cwd?: string,
    silent = false
  ): Promise<string> {
    if (!args.length) {
      throw new Error("Command args must not be empty");
    }

    if (cwd && !existsSync(cwd)) {
      throw new Error(`Directory "${cwd}" does not exist`);
    }

    let cmd: string;
    if (Array.isArray(args)) {
      cmd = args.shift() || "";
    } else {
      cmd = args;
      args = [];
    }

    return new Promise((resolve, reject) => {
      const child = spawn(cmd, args as string[], {
        stdio: silent ? "pipe" : "inherit",
        shell: true,
        windowsVerbatimArguments: true,
        cwd,
      });

      let output = "";
      let error = "";

      child.on("exit", function (code) {
        if (code) {
          return reject(error);
        }
        resolve(output);
      });

      if (child.stdout) {
        child.stdout.on("data", (data) => {
          output += `\n${data}`;
        });
      }
      if (child.stderr) {
        child.stderr.on("data", (data) => {
          error += `\n${data}`;
        });
      }
    });
  }
}
