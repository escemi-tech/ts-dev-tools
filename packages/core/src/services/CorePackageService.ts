import { resolve } from "path";

import { PackageJson } from "./PackageJson";

export class CorePackageService {
  static getPackageRootPath() {
    return resolve(__dirname, "../..");
  }

  static getPackageName() {
    return PackageJson.fromDirPath(CorePackageService.getPackageRootPath()).getPackageName();
  }
}
