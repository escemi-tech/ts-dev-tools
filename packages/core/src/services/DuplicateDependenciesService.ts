import { PackageJson } from "./PackageJson";
import { Plugin, PluginService } from "./PluginService";

type DuplicateDependencies = Map<string, Set<string>>;

export class DuplicateDependenciesService {
  static executeDeduplication(absoluteProjectDir: string): void {
    console.info(`Checking for duplicate dev dependencies...`);

    const duplicateDependencies: DuplicateDependencies = new Map();
    DuplicateDependenciesService.getProjectDuplicateDependencies(
      absoluteProjectDir,
      duplicateDependencies
    );

    DuplicateDependenciesService.printDuplicatedDependencies(duplicateDependencies);

    console.info(`Check for duplicate dev dependencies done!`);
  }

  private static getProjectDuplicateDependencies(
    absoluteProjectDir: string,
    duplicateDependencies: DuplicateDependencies
  ) {
    const installedPlugins = PluginService.getInstalledPlugins(absoluteProjectDir);

    const projectPackageJson = PackageJson.fromDirPath(absoluteProjectDir);
    const projectDevDependencies = projectPackageJson.getDevDependenciesPackageNames();

    for (const plugin of installedPlugins) {
      DuplicateDependenciesService.getPluginDuplicateDependencies(
        plugin,
        projectDevDependencies,
        duplicateDependencies
      );
    }

    return duplicateDependencies;
  }

  private static getPluginDuplicateDependencies(
    plugin: Plugin,
    projectDevDependencies: string[],
    duplicateDependencies: DuplicateDependencies
  ) {
    // First check for duplicate of inherited plugins
    DuplicateDependenciesService.getProjectDuplicateDependencies(
      plugin.path,
      duplicateDependencies
    );

    const pluginPackageJson = PackageJson.fromDirPath(plugin.path);
    const pluginDependencies = pluginPackageJson.getDependenciesPackageNames();

    let pluginDuplicateDependencies = duplicateDependencies.get(plugin.fullname);
    if (!pluginDuplicateDependencies) {
      pluginDuplicateDependencies = new Set();
      duplicateDependencies.set(plugin.fullname, pluginDuplicateDependencies);
    }

    for (const projectDevDependency of projectDevDependencies) {
      if (pluginDependencies.includes(projectDevDependency)) {
        pluginDuplicateDependencies.add(projectDevDependency);
      }
    }
  }

  private static printDuplicatedDependencies(duplicateDependencies: DuplicateDependencies) {
    let hasDuplicates = false;
    duplicateDependencies.forEach((pluginDuplicateDependencies, plugin) => {
      if (pluginDuplicateDependencies.size) {
        const pluginDuplicateDependenciesValue = Array.from(pluginDuplicateDependencies);
        hasDuplicates = true;
        console.info(
          `Some dev dependencies are unnecessarily installed as their are already required by "${plugin}":\n  - ${pluginDuplicateDependenciesValue.join(
            "\n  - "
          )}\n`
        );
      }
    });

    if (!hasDuplicates) {
      console.info(`No duplicate dev dependencies found.`);
    }
  }
}
