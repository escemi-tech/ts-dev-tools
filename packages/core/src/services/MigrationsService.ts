import { readdirSync, unlinkSync } from "fs";
import { resolve } from "path";

import { PackageJson } from "../services/PackageJson";
import { Plugin, PluginService } from "./PluginService";

export type Migration = { fullname: string; shortname: string; path: string };

export type MigrationUpFunction = (absoluteProjectDir: string) => Promise<void>;

export class MigrationsService {
  static MIGRATION_BUILT_PATH = "dist/install/migrations";

  static async executeMigrations(
    absoluteProjectDir: string,
    currentVersion: string | undefined
  ): Promise<void> {
    const migrations = MigrationsService.getAvailableMigrations(absoluteProjectDir, currentVersion);

    const packageJson = PackageJson.fromDirPath(absoluteProjectDir);
    const packageJsonBackupPath = packageJson.backup();

    try {
      for (const migration of migrations) {
        console.info(`Applying migration "${migration.fullname}"...`);

        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const { up } = require(migration.path);

        // Apply migration
        await up(absoluteProjectDir);

        // Upgrade current version
        PackageJson.fromDirPath(absoluteProjectDir).merge({
          tsDevTools: { version: migration.shortname },
        });

        console.info(`Migration "${migration.fullname}" applied!`);
      }
    } catch (error) {
      // Rollback package.json
      packageJson.restore(packageJsonBackupPath);
      throw error;
    }

    unlinkSync(packageJsonBackupPath);
  }

  private static getAvailableMigrations(
    absoluteProjectDir: string,
    currentVersion: string | undefined
  ): Migration[] {
    const installedPlugins = PluginService.getInstalledPlugins(absoluteProjectDir);

    const migrationFiles: Migration[] = [];
    for (const installedPlugin of installedPlugins) {
      migrationFiles.push(
        ...MigrationsService.getPluginMigrations(installedPlugin, currentVersion)
      );
    }

    migrationFiles.sort(({ shortname: nameA }, { shortname: nameB }) => nameA.localeCompare(nameB));

    return Array.from(new Set(migrationFiles));
  }

  private static getPluginMigrations(
    plugin: Plugin,
    currentVersion: string | undefined
  ): Migration[] {
    const migrationFiles = [];

    const pluginMigrationsDirPath = resolve(plugin.path, MigrationsService.MIGRATION_BUILT_PATH);
    for (const migrationFile of readdirSync(pluginMigrationsDirPath)) {
      if (!migrationFile.match(/^[0-9]{14}-[-a-z]+\.(js|ts)$/)) {
        continue;
      }

      const migrationName = MigrationsService.getMigrationNameFromFile(migrationFile);
      const shouldApplyMigration = MigrationsService.migrationIsAfterCurrentVersion(
        migrationName,
        currentVersion
      );

      if (!shouldApplyMigration) {
        continue;
      }

      migrationFiles.push({
        shortname: migrationName,
        fullname: `${plugin.shortname} - ${migrationName}`,
        path: resolve(pluginMigrationsDirPath, migrationFile),
      });
    }
    return migrationFiles;
  }

  public static getMigrationNameFromFile(migrationFile: string): string {
    const migrationName = migrationFile.split(".").slice(0, -1).join(".");
    return migrationName;
  }

  public static migrationIsAfterCurrentVersion(migrationName: string, currentVersion?: string) {
    return !currentVersion || currentVersion.localeCompare(migrationName) < 0;
  }
}
