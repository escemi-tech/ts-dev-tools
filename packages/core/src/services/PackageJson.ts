import { PackageJson as PackageJsonType } from "type-fest";

import { PackageJsonMerge } from "./PackageJsonMerge";
import { FileService } from "./FileService";
import { join } from "path";

export type JsonArray = boolean[] | number[] | string[] | JsonFileData[] | Date[];
export type AnyJson = boolean | number | string | JsonFileData | Date | JsonArray | JsonArray[];

export type JsonFileData = {
  [key: string]: AnyJson | undefined;
};
export type PackageJsonContent = JsonFileData & PackageJsonType;

export class PackageJson {
  private static readonly PACKAGE_JSON_FILE_NAME = "package.json";

  private content?: PackageJsonContent = undefined;

  constructor(private path: string) {
    if (!FileService.fileExists(this.path)) {
      throw new Error(`Package.json "${this.path}" does not exist`);
    }
  }

  getPath(): string {
    return this.path;
  }

  getContent(): PackageJsonContent {
    if (this.content) {
      return this.content;
    }
    return (this.content = JSON.parse(FileService.getFileContent(this.path)) as PackageJsonContent);
  }

  setContent(content: PackageJsonContent): void {
    this.content = content;
    this.write();
  }

  getPackageName(): string | undefined {
    return this.getContent().name;
  }

  getPackageVersion(): string | undefined {
    return this.getContent().version;
  }

  isPrivate(): boolean {
    return this.getContent().private === true;
  }

  getTsDevToolsVersion(): string | undefined {
    const tsDevToolsConfig = this.getContent().tsDevTools as JsonFileData | undefined;
    const version = tsDevToolsConfig?.version as string | undefined;
    return version;
  }

  getDependenciesPackageNames(): string[] {
    const dependencies = this.getContent().dependencies;
    return dependencies ? Object.keys(dependencies) : [];
  }

  getDevDependenciesPackageNames(): string[] {
    const devDependencies = this.getContent().devDependencies;
    return devDependencies ? Object.keys(devDependencies) : [];
  }

  getAllDependenciesPackageNames(): string[] {
    return Array.from(
      new Set([...this.getDependenciesPackageNames(), ...this.getDevDependenciesPackageNames()])
    );
  }

  merge(update: PackageJsonContent): void {
    this.content = PackageJsonMerge.merge(this.getContent(), update);
    this.write();
  }

  backup(): string {
    const backupPath = this.path + ".backup";
    FileService.copyFile(this.path, backupPath);
    return backupPath;
  }

  restore(backupPath: string): void {
    FileService.copyFile(backupPath, this.path);
    this.content = undefined;
  }

  private write() {
    FileService.putFileContent(this.path, JSON.stringify(this.content, null, 2));
  }

  static fromDirPath(dirPath: string): PackageJson {
    const packageJsonPath = join(dirPath, PackageJson.PACKAGE_JSON_FILE_NAME);
    return new PackageJson(packageJsonPath);
  }
}
