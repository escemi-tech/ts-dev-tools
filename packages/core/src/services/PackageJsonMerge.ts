import { AnyJson, JsonArray, JsonFileData, PackageJsonContent } from "./PackageJson";

export class PackageJsonMerge {
  static merge(source: PackageJsonContent, update: PackageJsonContent): PackageJsonContent {
    return PackageJsonMerge.mergeObjects(source, update);
  }

  private static mergeValues(
    source: AnyJson | undefined,
    update: AnyJson | undefined
  ): AnyJson | undefined {
    if (source === undefined) {
      return update;
    }

    if (!PackageJsonMerge.typesEqual(source, update)) {
      throw new Error(`Unable to merge package json value because types are different`);
    }

    // Deal with arrays
    if (Array.isArray(update)) {
      return PackageJsonMerge.mergeArrays(source as JsonArray[], update as JsonArray[]);
    }

    // Deal with objects
    if (typeof update === "object") {
      return PackageJsonMerge.mergeObjects(
        source as PackageJsonContent,
        update as PackageJsonContent
      );
    }

    return update;
  }

  private static mergeObjects(source: JsonFileData, update: JsonFileData): JsonFileData {
    for (const updateKey in update) {
      if (!Object.prototype.hasOwnProperty.call(update, updateKey)) {
        continue;
      }
      source[updateKey] = PackageJsonMerge.mergeValues(source[updateKey], update[updateKey]);
    }
    return source;
  }

  private static mergeArrays(source: JsonArray[], update: JsonArray[]): JsonArray[] {
    for (const item of update) {
      if (!source.includes(item)) {
        source.push(item);
      }
    }
    return source;
  }

  private static typesEqual(sourceValue: unknown, updateValue: unknown) {
    if (Array.isArray(sourceValue) && !Array.isArray(updateValue)) {
      return false;
    }

    return typeof sourceValue === typeof updateValue;
  }
}
