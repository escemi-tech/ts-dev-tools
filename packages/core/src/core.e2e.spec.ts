import { resolve } from "path";

import { PackageJson } from "./services/PackageJson";
import { exec, safeExec } from "./tests/cli";
import { createProjectForTestFile, deleteTestProject } from "./tests/test-project";
import { createTestMonorepoProject } from "./tests/test-project-monorepo";
import { createTestPackagesDir } from "./tests/test-packages";
import { deleteFolderRecursive } from "./tests/file-system";
import { stripAnsi } from "./tests/console";

// Set to false to avoid using the cache
const useCache = true;
// Set to false to inspect the test project directory after the test
const shouldCleanupAfterTest = true;

async function typescriptProjectGenerator(projectDir: string) {
  await safeExec(projectDir, "npm create vite . -- --template vanilla-ts");
  await safeExec(projectDir, "npm install");
}

function removeInstallWarnings(installPackageStderr: string, packageToInstall: string): string {
  return installPackageStderr
    .split("\n")
    .filter((line) => {
      const lineWithoutColors = stripAnsi(line);
      return (
        !lineWithoutColors.includes(`warning "${packageToInstall}`) &&
        !lineWithoutColors.includes(`warning ${packageToInstall}`)
      );
    })
    .join("\n");
}

const packageToTest = "core";
describe(`E2E - ${packageToTest}`, () => {
  let testProjectDirPackages: string;
  let packageToInstall: string;

  beforeAll(async () => {
    if (!useCache) {
      console.warn("Cache is disabled. Enable it one dev is done.");
    }
    if (!shouldCleanupAfterTest) {
      console.warn("Cleanup is disabled. Enable it one dev is done.");
    }

    testProjectDirPackages = await createTestPackagesDir(__filename);
    const packagePath = resolve(testProjectDirPackages, "packages", packageToTest);
    packageToInstall = `${packagePath}`;
  }, 200000);

  afterEach(async () => {
    if (shouldCleanupAfterTest) {
      await deleteTestProject(__filename);
    }
  });

  afterAll(async () => {
    if (shouldCleanupAfterTest) {
      await deleteFolderRecursive(testProjectDirPackages);
    }
  });

  describe("Simple project", () => {
    let testProjectDir: string;

    beforeEach(async () => {
      testProjectDir = await createProjectForTestFile(
        __filename,
        useCache,
        typescriptProjectGenerator
      );
    }, 200000);

    it(`Installs ${packageToTest} package`, async () => {
      const { code: installPackageCode, stderr: installPackageStderr } = await exec(
        testProjectDir,
        `npm install --save-dev "${packageToInstall}"`
      );

      expect(removeInstallWarnings(installPackageStderr, packageToInstall)).toBeFalsy();
      expect(installPackageCode).toBe(0);

      const {
        code: installCode,
        stderr: stderrCode,
        stdout,
      } = await exec(testProjectDir, "npm exec ts-dev-tools install");

      expect(stripAnsi(stdout)).toMatchSnapshot();
      expect(stderrCode).toBeFalsy();
      expect(installCode).toBe(0);

      const packageJson = PackageJson.fromDirPath(testProjectDir);
      expect(packageJson.getTsDevToolsVersion()).not.toBeFalsy();
      expect(packageJson.getContent().scripts).toMatchSnapshot();
      expect(packageJson.getContent().prettier).toMatchSnapshot();
      expect(packageJson.getContent().jest).toMatchSnapshot();

      const { code: lintCode, stderr: lintStderr } = await exec(testProjectDir, "npm run lint");

      expect(lintStderr).toBeFalsy();
      expect(lintCode).toBe(0);

      const { code: buildCode, stderr: buildStderr } = await exec(testProjectDir, "npm run build");

      expect(buildStderr).toBeFalsy();
      expect(buildCode).toBe(0);
    }, 200000);
  });

  describe("Monorepo project", () => {
    let testProjectDir: string;

    beforeEach(async () => {
      testProjectDir = await createTestMonorepoProject(
        __filename,
        useCache,
        typescriptProjectGenerator
      );
    }, 200000);

    it(`Installs ${packageToTest} package`, async () => {
      const { code: installPackageCode, stderr: installPackageStderr } = await exec(
        testProjectDir,
        `npm install --save-dev -W "${packageToInstall}"`
      );

      expect(removeInstallWarnings(installPackageStderr, packageToInstall)).toBeFalsy();
      expect(installPackageCode).toBe(0);

      const {
        code: installCode,
        stderr: stderrCode,
        stdout,
      } = await exec(testProjectDir, "npm exec ts-dev-tools install");

      expect(stripAnsi(stdout)).toMatchSnapshot();
      expect(stderrCode).toBeFalsy();
      expect(installCode).toBe(0);

      const packageJson = PackageJson.fromDirPath(testProjectDir);
      expect(packageJson.getTsDevToolsVersion()).not.toBeFalsy();
    }, 200000);
  });
});
