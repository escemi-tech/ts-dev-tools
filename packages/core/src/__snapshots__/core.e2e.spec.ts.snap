// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E - core Monorepo project Installs core package 1`] = `
"Installing ts-dev-tools...
Applying migration "core - 20201024173398-init"...
Migration "core - 20201024173398-init" applied!
Applying migration "core - 20220617100200-prettier-cache"...
Migration "core - 20220617100200-prettier-cache" applied!
Applying migration "core - 20240329200200-eslint-ignore"...
Migration "core - 20240329200200-eslint-ignore" applied!
Applying migration "core - 20240412185500-eslint-config"...
Migration "core - 20240412185500-eslint-config" applied!
Applying migration "core - 20240617094000-config-nx-scopes"...
Migration "core - 20240617094000-config-nx-scopes" applied!
Symlinking dev dependencies...
- Symlinking @commitlint/cli
- Symlinking @commitlint/config-conventional
- Symlinking @eslint/js
- Symlinking @types/jest
- Symlinking eslint
- Symlinking eslint-config-prettier
- Symlinking eslint-plugin-jest
- Symlinking import-sort-style-module
- Symlinking jest
- Symlinking lint-staged
- Symlinking prettier
- Symlinking prettier-plugin-import-sort
- Symlinking pretty-quick
- Symlinking ts-jest
- Symlinking typescript-eslint
Symlinking dev dependencies done!
Checking for duplicate dev dependencies...
Some dev dependencies are unnecessarily installed as their are already required by "@ts-dev-tools/core":
  - typescript

Check for duplicate dev dependencies done!
Installation done!"
`;

exports[`E2E - core Simple project Installs core package 1`] = `
"Installing ts-dev-tools...
Applying migration "core - 20201024173398-init"...
Migration "core - 20201024173398-init" applied!
Applying migration "core - 20220617100200-prettier-cache"...
Migration "core - 20220617100200-prettier-cache" applied!
Applying migration "core - 20240329200200-eslint-ignore"...
Migration "core - 20240329200200-eslint-ignore" applied!
Applying migration "core - 20240412185500-eslint-config"...
Migration "core - 20240412185500-eslint-config" applied!
Applying migration "core - 20240617094000-config-nx-scopes"...
Migration "core - 20240617094000-config-nx-scopes" applied!
Symlinking dev dependencies...
- Symlinking @commitlint/cli
- Symlinking @commitlint/config-conventional
- Symlinking @eslint/js
- Symlinking @types/jest
- Symlinking @types/node
- Symlinking eslint
- Symlinking eslint-config-prettier
- Symlinking eslint-plugin-jest
- Symlinking import-sort-style-module
- Symlinking jest
- Symlinking lint-staged
- Symlinking prettier
- Symlinking prettier-plugin-import-sort
- Symlinking pretty-quick
- Symlinking ts-jest
- Symlinking typescript-eslint
Symlinking dev dependencies done!
Checking for duplicate dev dependencies...
Some dev dependencies are unnecessarily installed as their are already required by "@ts-dev-tools/core":
  - typescript

Check for duplicate dev dependencies done!
Installation done!"
`;

exports[`E2E - core Simple project Installs core package 2`] = `
{
  "build": "tsc --noEmit",
  "dev": "vite",
  "format": "prettier --cache --write '**/*.ts'",
  "jest": "jest --detectOpenHandles --forceExit",
  "lint": "eslint "src/**/*.{ts,tsx}"",
  "prepare": "ts-dev-tools install",
  "preview": "vite preview",
  "test": "npm run jest --maxWorkers=50%",
  "test:ci": "npm run test:cov --runInBand",
  "test:cov": "npm run test --coverage",
  "test:watch": "npm run jest --watch --maxWorkers=25%",
}
`;
